src/bokeh/util/hex.py

```diff
@@ -46,12 +46,29 @@
     'axial_to_cartesian',
     'cartesian_to_axial',
     'hexbin',
+    'HexBinResult',
 )
 
 #-----------------------------------------------------------------------------
 # General API
 #-----------------------------------------------------------------------------
 
+
+@dataclass
+class HexBinResult:
+    q: npt.NDArray[np.integer]
+    r: npt.NDArray[np.integer]
+    counts: npt.NDArray[np.integer]
+
+    def __getitem__(self, key: str) -> npt.NDArray[np.integer]:
```

Comment:
We could maybe do something with this type hint so it warns the user.

Reply:
This could be done with something like this:

```suggestion
    def __getitem__(self, key: Never) -> Never:
```
``` bash
example_type.py:10:6: error: Invalid index type "str" for "HexBinResult";
expected type "Never"  [index]
    bins["q"]
         ^~~
Found 1 error in 1 file (checked 1 source file)
```

``` python
import numpy as np

from bokeh.util.hex import hexbin

n = 50000
x = np.random.standard_normal(n)
y = np.random.standard_normal(n)

bins = hexbin(x, y, 0.1)
bins["q"]
```

Reply:
Just clarifying: you want a type checker to error if a bad literal key is used? Would `Literal["q", "r", "counts"]` not suffice? Would retain the runtime check, regardless.

Reply:
I wanted it to raise a type check error if any getitem was used. So it could be removed at some point, as it is mostly a compatibility layer to mimic the DataFrame API. 

But this was just a suggestion.

Reply:
Eh, this seems like one of those things we don't make a point to advertise, but also just keep around for simplicity. If we really want to remove it, then 4.0 would be reasonable.

Reply:
Using `Never` type actually implies that something is impossible to happen, like when an exception is thrown then a function cannot return any value, thus its return type is `Never`. In this case we want this function to accept certain values and produce output values. A deprecation warning would be in order, i.e. `warnings.deprecated()` (PEP702), which works with the type system.

Reply:
>  A deprecation warning would be in order, i.e. warnings.deprecated() (PEP702), which works with the type system.

This was my first attempt, but I couldn't get it to work with `__getitem__`. 

> Using Never type actually implies that something is impossible to happen

I wanted to generate an equivalent type-checking error as if `__getitem__` did not exist, while still maintaining it as a runtime behavior. 

> Eh, this seems like one of those things we don't make a point to advertise

Didn't want to advertise this at all. Adding `__getitem__` is for runtime backward compatibility.

Reply:
Indeed `warnings.deprecated()` doesn't currently show up, at least, in vscode, though it works for methods, etc. That's unfortunate, but maybe it's either a bug or a TODO.